### PR TITLE
Allow disabling ``duplicate-code`` with a disable comment

### LINKED ISSUE #214: The duplicate-code (R0801) can't be disabled
Originally reported by: **Anonymous**

---

It's seems like it's not possible to disable the duplicate code check on portions of a file. Looking at the source, I can see why as it's not a trivial thing to do (if you want to maintain the same scope semantics as other #pylint:enable/disable comments. This would be nice to have though (or I guess I could just cleanup my duplicate code).

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/214

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

**Please read**

The most important reason why nobody could get this to work (I think) is because the similarity check can be ran stand-alone. I'm not sure what the rationale for this is, but I think that might have caused issues for other contributors.

This PR checks whether `self` has `linter` attribute (and therefore is not being run standalone) and then checks for disables. I think this is a fair approach to the problem: if you're running the similarity checker on its own it shouldn't be disabled by `pylint` type comments. 

I would like the opinion of maintainers though as looking at the issues and past PRs has shown that this discussion has been quite opinionated.

Potentially closes #214

### CODE REVIEW

**Review by earonesty — APPROVED** [review]

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas After your merge commit this is now changing 40 files? Are you sure you merged the right `main` branch?

**Comment by DanielNoord** [comment]
Good to see `pydocstringformatter` is working as expected 😄

**Comment by Pierre-Sassoulas** [comment]
> This PR checks whether self has linter attribute (and therefore is not being run standalone) and then checks for disables. I think this is a fair approach to the problem: if you're running the similarity checker on its own it shouldn't be disabled by pylint type comments.

Is there no way to easily take the disable into account when launched as a standalone ? If not I think it's fair enough, being able to disable when run with pylint is one of the most asked feature so having this one is pretty important.

File: pylint/checkers/similar.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:388]: That's pretty clever.

**Comment by DanielNoord** [comment]
> > This PR checks whether self has linter attribute (and therefore is not being run standalone) and then checks for disables. I think this is a fair approach to the problem: if you're running the similarity checker on its own it shouldn't be disabled by pylint type comments.
> 
> Is there no way to easily take the disable into account when launched as a standalone ? If not I think it's fair enough, being able to disable when run with pylint is one of the most asked feature so having this one is pretty important.

Well, the issue with that is that the `disable` param checking is all done on the `PyLinter` class. In standalone mode no `PyLinter` class is instantiated so we would need to write a completely separate piece of code to handle this. I think this is why the previous attempts at this all failed: you need to duplicate a lot of work from `PyLinter` for this to be able to run in standalone and in `pylint` mode with the same form of disabling.

File: tests/test_similar.py
Comment (Pierre-Sassoulas) [inline tests/test_similar.py:59]: I feel maybe using ``pathlib.Path`` would permit to remove this but I'm not certain of what the ``re.sub`` is doing.

**Comment by Pierre-Sassoulas** [comment]
> In standalone mode no PyLinter class is instantiated so we would need to write a completely separate piece of code to handle this

Maybe we can instead instantiate a PyLinter with ``--disable=all --enable=duplicate-code`` ?

File: tests/test_similar.py
Comment (DanielNoord) [reply tests/test_similar.py:59]: It removes the `Users/.../.../.../pylint/` element from the `output` so that we get a direct path to the test file. This is because in normal `pylint` output you get the full path name and we don't want that there because the path names changes depending on test environment (`Users/Daniel/...` vs. `Users/Pierre/...`)

**Comment by DanielNoord** [comment]
> Maybe we can instead instantiate a PyLinter with `--disable=all --enable=duplicate-code` ?

Haven't looked into that, but that kind of defeats the purpose of having this be a stand-alone tool. You would then also get all the overhead of configuration loading etc.

Personally I think we should be fine with adding this and then even closing the issue.
If somebody really wants to disable the duplicate-code checker while they are running `symilar` stand-alone (why even run it in the first place then...) they can create a new issue and we can then discuss if there is a real use case for that.

File: tests/test_similar.py
Comment (Pierre-Sassoulas) [reply tests/test_similar.py:59]: He, then we can use [os.path.relpath()](https://docs.python.org/3/library/os.path.html#os.path.relpath) or [PurePath.relative_to()](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.relative_to).

**Comment by Pierre-Sassoulas** [comment]
>  (why even run it in the first place then...)

Right, it's possible to launch pylint with ``--disable=all --enable=duplicate-code``directly instead. And it's also possible to use a specialized tool like [Simian or PMD directly](https://stackoverflow.com/questions/191614/how-to-detect-code-duplication-during-development)

File: tests/test_similar.py
Comment (DanielNoord) [reply tests/test_similar.py:59]: Not really, since these aren't single strings.

```console
❯ pylint test_dir/another_dir/test_data/test.py
************* Module test
test_dir/another_dir/test_data/test.py:1:0: W0104: Statement seems to have no effect (pointless-statement)
test_dir/another_dir/test_data/test.py:2:0: W0104: Statement seems to have no effect (pointless-statement)
test_dir/another_dir/test_data/test.py:3:0: W0104: Statement seems to have no effect (pointless-statement)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

It is the output similar to this. However, depending on how `Pylint` is invoked the `Users/Daniel/...` is added as well. This normalises strings that refer to paths in the normal `pylint` output.

Comment (Pierre-Sassoulas) [reply tests/test_similar.py:59]: Hmm yeah, I remember that we had something similar in the config and had to use abspath and relpath in the result : https://github.com/PyCQA/pylint/blob/main/tests/config/test_functional_config_loading.py#L10

Comment (DanielNoord) [reply tests/test_similar.py:59]: We seem to handle it here:
https://github.com/PyCQA/pylint/blob/a1df7685a4e6a05b519ea011f16a2f0d49d08032/pylint/testutils/configuration_test.py#L92

I don't have a particular preference, but this seems to work fine and I don't think we will need to expand these tests much later on. So this hacky way of substituting using regex would be fine imo.

**Review by Pierre-Sassoulas — APPROVED** [review]
There's currently 121 👍 on the original issue, safe to say (1) you're going made a lot of person happy with this one 😄 

(1) unless a catastrophic failure condition eluded me during review 😂

**Comment by DanielNoord** [comment]
> There's currently 121 👍 on the original issue, safe to say (1) you're going made a lot of person happy with this one 😄
> 
> (1) unless a catastrophic failure condition eluded me during review 😂

Yeah I completely forgot about this, but my response saying that I had a potential fix also got 12 👍... I'll do one final review as this was quite long ago and then I'll merge (if that's okay with you of course!).

**Comment by DanielNoord** [comment]
Merging this! @earonesty also thanks for your review!

**Pull request merged**
